### PR TITLE
GH-34248: [C++] Add an order_by node

### LINKED ISSUE #34248: [C++] Add an order_by node which can reassign an ordering mid-plan
### Describe the enhancement requested

#34136 added ordering to exec plans.  We can now take our order_by_sink node and make an actual order_by node which can be placed anywhere in a plan.  This can use the same naive implementation in order_by_sink.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by westonpace**
* Closes: #34248

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
To review, this needs to be rebased first now https://github.com/apache/arrow/pull/34137 is merged?

**Comment by westonpace** [comment]
> To review, this needs to be rebased first now https://github.com/apache/arrow/pull/34137 is merged?

Yes, thanks.  I've done that now.

File: cpp/src/arrow/compute/exec/options.h [hunk 1]
```diff
@@ -402,6 +402,14 @@ class ARROW_EXPORT OrderBySinkNodeOptions : public SinkNodeOptions {
   SortOptions sort_options;
 };
 
+class ARROW_EXPORT OrderByNodeOptions : public ExecNodeOptions {
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/options.h:405]: nit: docstring?

**Review by lidavidm — APPROVED** [review]

File: cpp/src/arrow/compute/exec/options.h [hunk 2]
```diff
@@ -402,6 +402,14 @@ class ARROW_EXPORT OrderBySinkNodeOptions : public SinkNodeOptions {
...
+ public:
+  static constexpr std::string_view kName = "order_by";
+  explicit OrderByNodeOptions(Ordering ordering) : ordering(std::move(ordering)) {}
+
+  Ordering ordering;
```
[the remark this answers is not in the document]
Comment (westonpace) [reply cpp/src/arrow/compute/exec/options.h:410]: Yes, `SortOptions` and `Ordering` are equivalent.  However, `Ordering` doesn't extend from `FunctionOptions`.  I wasn't sure how to proceed here so I ended up creating a new type.  It felt a bit weird for a `FunctionOptions` to be used in exec plans completely unrelated to running functions.  Although, to play devil's advocate, `FunctionOptions` is "mostly" just a marker interface and doesn't really prevent anything from working so I would open to just using that.

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply cpp/src/arrow/compute/exec/options.h:410]: I think the Ordering certainly makes sense, and it's quite trivial to convert the one into the other, I was just wondering for a consistent API if we want to add a constructor version that accepts the SortOptions (or for OrderBySinkNode to add a version that accepts Ordering)

**Review by lidavidm — APPROVED** [review]

**Comment by westonpace** [comment]
Sorry it took so long to get back to this.  @jorisvandenbossche thanks for the testing and exploration.  For others reading this I should mention this conversation continued on Zulip as well.  The outcome was that we should maybe make the batch size configurable but also maybe use a larger batch size in general.

**Comment by jorisvandenbossche** [comment]
Thanks for finishing up! Pushed the python bindings I wrote to a new PR: https://github.com/apache/arrow/pull/34654

**Pull request merged**